src/Libraries/Microsoft.Extensions.Compliance.Redaction/HmacRedactor.cs

```diff
@@ -0,0 +1,152 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Security.Cryptography;
+using Microsoft.Extensions.Options;
+using Microsoft.Shared.Diagnostics;
+using Microsoft.Shared.Text;
+
+#if NET6_0_OR_GREATER
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+#else
+using System.Diagnostics.CodeAnalysis;
+using System.Text;
+#endif
+
+namespace Microsoft.Extensions.Compliance.Redaction;
+
+internal sealed class HmacRedactor : Redactor
+{
+#if NET6_0_OR_GREATER
+    private const int SHA256HashSizeInBytes = 32;
+#endif
+    private const int BytesOfHashWeUse = 16;
+
+    /// <remarks>
+    /// Magic numbers are formula for calculating base64 length with padding.
+    /// </remarks>
+    private const int Base64HashLength = ((BytesOfHashWeUse + 2) / 3) * 4;
+
+    private readonly int _redactedLength;
+    private readonly byte[] _hashKey;
+    private readonly string _keyId;
+
+    public HmacRedactor(IOptions<HmacRedactorOptions> options)
+    {
+        var value = Throw.IfMemberNull(options, options.Value);
+
+        _hashKey = Convert.FromBase64String(value.Key);
+        _keyId = value.KeyId.HasValue ? value.KeyId.Value.ToInvariantString() + ':' : string.Empty;
+        _redactedLength = Base64HashLength + _keyId.Length;
+    }
+
+    public override int GetRedactedLength(ReadOnlySpan<char> source)
+    {
+        if (source.IsEmpty)
+        {
+            return 0;
+        }
+
+        return _redactedLength;
+    }
+
+#if NET6_0_OR_GREATER
+    public override int Redact(ReadOnlySpan<char> source, Span<char> destination)
+    {
+        var length = GetRedactedLength(source);
+        if (length == 0)
+        {
+            return 0;
+        }
+
+        Throw.IfBufferTooSmall(destination.Length, length, nameof(destination));
+
+        _keyId.AsSpan().CopyTo(destination);
+        return CreateSha256Hash(source, destination[_keyId.Length..], _hashKey) + _keyId.Length;
+    }
+
+    [SkipLocalsInit]
+    private static int CreateSha256Hash(ReadOnlySpan<char> source, Span<char> destination, byte[] hashKey)
+    {
+        Span<byte> hashBuffer = stackalloc byte[SHA256HashSizeInBytes];
+
+        _ = HMACSHA256.HashData(hashKey, MemoryMarshal.AsBytes(source), hashBuffer);
+        _ = Convert.TryToBase64Chars(hashBuffer.Slice(0, BytesOfHashWeUse), destination, out int charsWritten);
```

Comment:
What if `TryToBase64Chars()` returns `false`?

Reply:
Can't fail since we measured the size of the destination buffer before. I added a comment to that effect.